vignettes/data-repository.Rmd

```diff
@@ -0,0 +1,236 @@
+
+Although git can store binary files, its real power and efficiency is with plain text files. R has several functions to read and write `data.frames` as plain text files. Main downside is the loss of metadata, especially in case of `factors`. A `factor` will be written as a `character` and hence lose all information on the levels. This can be important when the levels are a) not sorted alphabetically, b) not all levels have observations or c) the `factor` is ordered. Besides that, storing a `factor` as a `character` is not efficient.
+
+Git stores the version history under the form of diffs: a list of lines which are deleted and a list of lines which are inserted at a specific line number in a file. We have to keep this in mind when storing data as plain text files in Git. The table below indicates the impact of changing the information content on the diff in Git. Note that changing observations requires a small diff, while changing variables results in a massive diff.
+
+
+|  Change in data | Git diff |
+|:----------------------|:----------|
+| remove 1 observation | remove 1 line |
+| add 1 observation | add 1 line |
+| update 1 observation | remove 1 line and add 1 line |
+| remove 1 variable | remove all lines and add all lines |
+| add 1 variable | remove all lines and add all lines |
+Table: Changes in the information content of the data
+
+In the next table we show the effect on the diff when making changes in the data which don't change the information content of the data.
+
+
+|Change in data     |Git diff                           |
+|:------------------|:----------------------------------|
+|move 1 observation |remove 1 line and add 1 line       |
+|move 1 variable    |remove all lines and add all lines |
+Table: Changes in the data without changing the information content
+
+So in order to use Git as a performant and efficient version control system for data, we need to make sure to store the metadata and keep the diffs as small as possible. In the next section, the way git2r handles this will be explained.
+
+## Tools in `git2r`
+
+### Connect to a repository
+
+```{r}
+# Create a directory in tempdir
+path <- tempfile(pattern = "git2r-")
+dir.create(path)
+
+library(git2r)
+# inititialize a new git repository
+repo <- init(path)
+config(repo, user.name = "Alice", user.email = "alice@example.org")
+```
+
+```{r eval = FALSE}
+# connect to an existing repository
+repo <- repository(path)
+```
+
+### Storing data
+
+Use `write_delim_git()` to store a `data.frame` into the repository. The function will separate the data and the metadata. The data is stored as a headerless, unquoted tab delimited file with ".tsv" extension and UTF-8 encoding. The metadata is stored in YAML format with ".yml" extension. Therefore any extension given to the `file` will be stripped (with a warning).
+
```

Comment:
Just to check if I get it right... the metadata contains the variable names, their type and in case of factors the mapping between levels and labels + the sorting order?

Should attributes of the data.frame be saved as metadata as well? For instance rownames?

Reply:
IMHO: if rownames are important they should be a variable